File: api/status.py
from __future__ import annotations

import json
import os
import urllib.request
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _human_eta(hours: float | None) -> str:
    if hours is None or hours < 0:
        return "—"
    if hours < 1:
        return f"~{int(hours * 60)}m"
    if hours < 48:
        return f"~{hours:.1f}h"
    return f"~{hours / 24.0:.1f}d"


def _normalize_upstream(data: dict) -> dict:
    """
    Normalize multiple possible upstream schemas into the dashboard schema expected by index.html:
      - ok, active, run_id, generated_at, server_time
      - progress: { completion_percent, towers_collected, total_tiles, completed_tiles, pending_tiles }
      - eta: { eta_hours, eta_human, projected_finish_at }
      - health: { state }  (kept simple for UI mapping)
    """
    # Already in our native schema?
    if "generated_at" in data and "progress" in data and "eta" in data:
        # Ensure required keys exist / are safe.
        data.pop("debug", None)
        data["workers"] = []
        return data

    # Example alternate schema (your Hetzner API):
    # {
    #   "timestamp": "...Z",
    #   "progress": {"total_completed_tiles": 123, "total_tiles": 456, "overall_progress_pct": 27.0},
    #   "towers": {"total": 215000},
    #   "estimate": {"remaining_hours": 100, "remaining_days": 4.2},
    #   "status": "running"
    # }
    ts = str(data.get("timestamp") or data.get("generated_at") or "") or None
    status = str(data.get("status") or "").lower().strip()
    active = status in ("running", "live", "active")

    prog = data.get("progress") if isinstance(data.get("progress"), dict) else {}
    total_tiles = prog.get("total_tiles", None)
    completed_tiles = prog.get("total_completed_tiles", None)
    pct = prog.get("overall_progress_pct", None)
    if pct is None and isinstance(total_tiles, (int, float)) and total_tiles:
        if isinstance(completed_tiles, (int, float)):
            pct = float(completed_tiles) / float(total_tiles) * 100.0

    towers = data.get("towers") if isinstance(data.get("towers"), dict) else {}
    towers_total = towers.get("total", None)

    est = data.get("estimate") if isinstance(data.get("estimate"), dict) else {}
    remaining_hours = est.get("remaining_hours", None)

    now_iso = _utc_now_iso()
    projected_finish_at = None
    if isinstance(remaining_hours, (int, float)) and remaining_hours >= 0:
        projected_finish_at = datetime.now(timezone.utc).timestamp() + float(remaining_hours) * 3600.0
        projected_finish_at = datetime.fromtimestamp(projected_finish_at, tz=timezone.utc).isoformat()

    # If upstream gives percent but not totals, keep totals null.
    pending_tiles = None
    if isinstance(total_tiles, (int, float)) and isinstance(completed_tiles, (int, float)):
        pending_tiles = max(0, int(total_tiles) - int(completed_tiles))

    health_state = "good" if active else "idle"

    return {
        "ok": True,
        "active": bool(active),
        "run_id": "live",
        "generated_at": ts or now_iso,
        "server_time": now_iso,
        "progress": {
            "total_tiles": int(total_tiles) if isinstance(total_tiles, (int, float)) else None,
            "completed_tiles": int(completed_tiles) if isinstance(completed_tiles, (int, float)) else None,
            "pending_tiles": pending_tiles,
            "deferred_tiles": None,
            "completion_percent": float(pct) if isinstance(pct, (int, float)) else None,
            "towers_collected": int(towers_total) if isinstance(towers_total, (int, float)) else 0,
        },
        "velocity": {"tiles_per_hour": None},
        "health": {"state": health_state},
        "eta": {
            "eta_hours": float(remaining_hours) if isinstance(remaining_hours, (int, float)) else None,
            "eta_human": _human_eta(float(remaining_hours)) if isinstance(remaining_hours, (int, float)) else "—",
            "projected_finish_at": projected_finish_at,
        },
        "workers": [],
        "debug": None,
    }


class handler(BaseHTTPRequestHandler):
    """
    Vercel Python Function entrypoint.

    This uses the supported BaseHTTPRequestHandler pattern. The frontend calls:
      GET /api/status

    Env vars (set in Vercel):
      - SCRAPER_STATUS_URL or CELLMAPPER_API_URL
      - SCRAPER_STATUS_BEARER_TOKEN or CELLMAPPER_API_TOKEN
      - ALLOW_INSECURE_STATUS_URL ("true" to allow http://)
    """

    def _send_json(self, status: int, payload: dict) -> None:
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Cache-Control", "no-store")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self) -> None:
        # Support both naming conventions.
        url = (os.environ.get("SCRAPER_STATUS_URL") or os.environ.get("CELLMAPPER_API_URL", "")).strip()
        token = (os.environ.get("SCRAPER_STATUS_BEARER_TOKEN") or os.environ.get("CELLMAPPER_API_TOKEN", "")).strip()
        allow_insecure = os.environ.get("ALLOW_INSECURE_STATUS_URL", "false").lower() == "true"

        if not url:
            self._send_json(
                500,
                {"ok": False, "error": "missing_env", "env": "SCRAPER_STATUS_URL|CELLMAPPER_API_URL"},
            )
            return
        if not token:
            self._send_json(
                500,
                {"ok": False, "error": "missing_env", "env": "SCRAPER_STATUS_BEARER_TOKEN|CELLMAPPER_API_TOKEN"},
            )
            return

        parsed = urlparse(url)
        if parsed.scheme not in ("https", "http") or not parsed.netloc:
            self._send_json(400, {"ok": False, "error": "invalid_url"})
            return
        if parsed.scheme == "http" and not allow_insecure:
            self._send_json(400, {"ok": False, "error": "insecure_url_not_allowed"})
            return

        req = urllib.request.Request(url, method="GET")
        req.add_header("Accept", "application/json")
        req.add_header("Authorization", f"Bearer {token}")

        try:
            with urllib.request.urlopen(req, timeout=10) as resp:
                raw = resp.read().decode("utf-8", errors="replace")
                data = json.loads(raw)
        except Exception:
            self._send_json(502, {"ok": False, "error": "upstream_unreachable"})
            return

        if not isinstance(data, dict):
            self._send_json(502, {"ok": False, "error": "invalid_upstream_payload"})
            return

        normalized = _normalize_upstream(data)
        normalized["workers"] = []
        normalized.pop("debug", None)
        self._send_json(200, normalized)

File: api/status.js
/**
 * Vercel Serverless Function (Node.js) – lightweight proxy to the scraper status endpoint.
 *
 * Env vars (set in Vercel):
 *   - SCRAPER_STATUS_URL or CELLMAPPER_API_URL
 *   - SCRAPER_STATUS_BEARER_TOKEN or CELLMAPPER_API_TOKEN
 *   - ALLOW_INSECURE_STATUS_URL = "true" to allow http:// (default: false)
 */

function utcNowIso() {
  return new Date().toISOString();
}

const FINISH_BUFFER_DAYS = 6;
const FINISH_BUFFER_MS = FINISH_BUFFER_DAYS * 24 * 60 * 60 * 1000;

function applyExpectedTotals(normalized) {
  const expectedTotal = Number.parseInt(process.env.EXPECTED_TOTAL_TILES || "", 10);
  if (!Number.isFinite(expectedTotal) || expectedTotal <= 0) return normalized;
  if (!normalized || typeof normalized !== "object" || !normalized.progress) return normalized;

  const out = { ...normalized, progress: { ...normalized.progress } };
  const completed = Number.isFinite(out.progress.completed_tiles) ? Number(out.progress.completed_tiles) : null;

  out.progress.total_tiles = expectedTotal;
  if (completed != null) {
    out.progress.pending_tiles = Math.max(0, expectedTotal - Math.trunc(completed));
    out.progress.completion_percent = expectedTotal > 0 ? (Number(completed) / expectedTotal) * 100 : null;
  } else {
    out.progress.pending_tiles = null;
    out.progress.completion_percent = null;
  }
  return out;
}

function humanEta(hours) {
  if (hours == null || !Number.isFinite(hours) || hours < 0) return "—";
  if (hours < 1) return `~${Math.max(0, Math.floor(hours * 60))}m`;
  if (hours < 48) return `~${hours.toFixed(1)}h`;
  return `~${(hours / 24).toFixed(1)}d`;
}

function normalizeUpstream(data) {
  // If upstream already matches our dashboard schema, pass through (with stripping).
  if (data && typeof data === "object" && data.generated_at && data.progress && data.eta) {
    const out = { ...data };
    delete out.debug;
    out.workers = [];

    // Apply +6 day buffer to projected finish (and keep ETA consistent).
    if (out.eta && typeof out.eta === "object") {
      if (out.eta.projected_finish_at) {
        const d = new Date(out.eta.projected_finish_at);
        if (!Number.isNaN(d.getTime())) {
          out.eta.projected_finish_at = new Date(d.getTime() + FINISH_BUFFER_MS).toISOString();
        }
      }
      if (Number.isFinite(out.eta.eta_hours)) {
        out.eta.eta_hours = Number(out.eta.eta_hours) + FINISH_BUFFER_DAYS * 24;
        out.eta.eta_human = humanEta(out.eta.eta_hours);
      }
    }

    return applyExpectedTotals(out);
  }

  const ts = (data && (data.timestamp || data.generated_at)) ? String(data.timestamp || data.generated_at) : null;
  const status = data && data.status ? String(data.status).toLowerCase().trim() : "";
  const active = ["running", "live", "active"].includes(status);

  const prog = data && typeof data.progress === "object" ? data.progress : {};
  const totalTiles = prog.total_tiles;
  const completedTiles = prog.total_completed_tiles;
  let pct = prog.overall_progress_pct;
  if (pct == null && Number.isFinite(totalTiles) && totalTiles && Number.isFinite(completedTiles)) {
    pct = (completedTiles / totalTiles) * 100;
  }

  const towers = data && typeof data.towers === "object" ? data.towers : {};
  const towersTotal = towers.total;

  const est = data && typeof data.estimate === "object" ? data.estimate : {};
  const remainingHours = est.remaining_hours;

  const nowIso = utcNowIso();
  let projectedFinishAt = null;
  if (Number.isFinite(remainingHours) && remainingHours >= 0) {
    projectedFinishAt = new Date(Date.now() + remainingHours * 3600 * 1000 + FINISH_BUFFER_MS).toISOString();
  }

  let pendingTiles = null;
  if (Number.isFinite(totalTiles) && Number.isFinite(completedTiles)) {
    pendingTiles = Math.max(0, Math.trunc(totalTiles) - Math.trunc(completedTiles));
  }

  return {
    ok: true,
    active: Boolean(active),
    run_id: "live",
    generated_at: ts || nowIso,
    server_time: nowIso,
    progress: {
      total_tiles: Number.isFinite(totalTiles) ? Math.trunc(totalTiles) : null,
      completed_tiles: Number.isFinite(completedTiles) ? Math.trunc(completedTiles) : null,
      pending_tiles: pendingTiles,
      deferred_tiles: null,
      completion_percent: Number.isFinite(pct) ? Number(pct) : null,
      towers_collected: Number.isFinite(towersTotal) ? Math.trunc(towersTotal) : 0,
    },
    velocity: { tiles_per_hour: null },
    health: { state: active ? "good" : "idle" },
    eta: {
      eta_hours: Number.isFinite(remainingHours) ? Number(remainingHours) + FINISH_BUFFER_DAYS * 24 : null,
      eta_human: Number.isFinite(remainingHours) ? humanEta(Number(remainingHours) + FINISH_BUFFER_DAYS * 24) : "—",
      projected_finish_at: projectedFinishAt,
    },
    workers: [],
    debug: null,
  };
}

module.exports = async (req, res) => {
  let stage = "start";
  try {
    const rawUrl =
      (process.env.SCRAPER_STATUS_URL || process.env.CELLMAPPER_API_URL || "").trim();
    const token =
      (process.env.SCRAPER_STATUS_BEARER_TOKEN || process.env.CELLMAPPER_API_TOKEN || "").trim();
    const allowInsecure = (process.env.ALLOW_INSECURE_STATUS_URL || "false").toLowerCase() === "true";
    const debug = (process.env.DEBUG_PROXY || "").toLowerCase() === "true";

    // Guard: people often paste env vars without newlines, causing concatenation.
    // Example: "http://.../api/statusCELLMAPPER_API_TOKEN=..."
    let url = rawUrl;
    const cutMarkers = [
      "CELLMAPPER_API_TOKEN=",
      "SCRAPER_STATUS_BEARER_TOKEN=",
      "SCRAPER_STATUS_URL=",
      "ALLOW_INSECURE_STATUS_URL=",
    ];
    for (const m of cutMarkers) {
      const idx = url.indexOf(m);
      if (idx > 0) {
        url = url.slice(0, idx).trim();
      }
    }

    if (!url) {
      res.status(500).json({ ok: false, error: "missing_env", env: "SCRAPER_STATUS_URL|CELLMAPPER_API_URL" });
      return;
    }
    if (!token) {
      res.status(500).json({ ok: false, error: "missing_env", env: "SCRAPER_STATUS_BEARER_TOKEN|CELLMAPPER_API_TOKEN" });
      return;
    }

    stage = "parse_url";
    let u;
    try {
      u = new URL(url);
    } catch {
      res.status(400).json({ ok: false, error: "invalid_url" });
      return;
    }
    if (u.protocol !== "https:" && u.protocol !== "http:") {
      res.status(400).json({ ok: false, error: "invalid_url" });
      return;
    }
    if (u.protocol === "http:" && !allowInsecure) {
      res.status(400).json({ ok: false, error: "insecure_url_not_allowed" });
      return;
    }

    stage = "fetch_upstream";
    const upstreamRes = await fetch(url, {
      method: "GET",
      headers: {
        Accept: "application/json",
        Authorization: `Bearer ${token}`,
      },
    });

    if (!upstreamRes.ok) {
      res.status(502).json({ ok: false, error: "upstream_unreachable", status: upstreamRes.status });
      return;
    }

    stage = "parse_upstream_json";
    let data;
    try {
      data = await upstreamRes.json();
    } catch {
      res.status(502).json({ ok: false, error: "invalid_upstream_json" });
      return;
    }
    if (!data || typeof data !== "object") {
      res.status(502).json({ ok: false, error: "invalid_upstream_payload" });
      return;
    }

    stage = "normalize";
    const normalized = applyExpectedTotals(normalizeUpstream(data));
    delete normalized.debug;
    normalized.workers = [];

    res.setHeader("Cache-Control", "no-store");
    res.status(200).json(normalized);
  } catch (e) {
    const debug = (process.env.DEBUG_PROXY || "").toLowerCase() === "true";
    res.status(502).json({
      ok: false,
      error: "proxy_error",
      stage,
      ...(debug ? { detail: String(e && e.message ? e.message : e) } : {}),
    });
  }
};


